### PR TITLE
New proxy server counting logic for agent and server

### PR DESCRIPTION
**Pull request by carreter**
An initial pass at the groundwork for #358 and #273 (i.e. allowing a dynamic proxy server count on both the server and the agent in a backwards-compatible manner). Currently a draft, depends on #632. Design doc [here](https://docs.google.com/document/d/16G8U-QWR4tNfY2jiOk-ZUEMVwBkq9R9iXZqXrxPsGOE/edit?usp=sharing).

### New package: `pkg/servercounter`
This package is structured around the `ServerCounter` interface:

```go
// A ServerCounter counts the number of available proxy servers.
type ServerCounter interface {
	CountServers() int
}
```

Currently provided are `StaticServerCounter`, which simply returns a stored int, and `CachedServerCounter`, which wraps another `ServerCounter` and provides catching functionality with a configurable cache expiry time.

### Proxy server
This PR wires the `ServerCounter` into the proxy server without changing its behavior by creating a `StaticServerCounter` with the command-line provided count.

### Proxy agent
This PR wires the `ServerCounter` into the proxy agent without changing its behavior by creating a `StaticServerCounter` with a default count of 0.

This PR also adds the `respectReceivedServerCount` field to the agent `ClientSet` struct. This field toggles whether the server count received from the proxy server should overwrite the proxy agent's server counter. Since I didn't want to change the behavior of the agent yet, this is set to `true` by default.

### Should `CountServers()` return an `error`?
Sometimes the `ServerCounter` will fail to get an updated server count. In this case, we could have `CountServers()` return `(int, error)` instead.

I opted against this because we can handle errors in the `ServerCounter` itself. This simplifies things on the caller side and allows for including a fallback server count in the `ServerCounter`.

### Should `ServerCounter` include an `UpdateCount(int)` method?
It might be useful for clients to (a) temporarily override the server count or (b) request an update to any cached server counts. They could do this by calling `UpdateCount()` with a non-negative value to temporarily override the count or with -1 to reset the override and request a count update.

I opted against this because the client can accomplish (a) using a `StaticServerCounter` (downside: this overwrites the previous `ServerCounter` if stored in the same field), and (b) seems like it would have limited use as the `ServerCounter` should be completely responsible for managing the server count.

A possible workaround is to add the `UpdateCount()` method to *just* `CachedServerCounter` rather than the entire interface. This allows for finer control of caching while maintaining a simple API for the general case.

### Should `respectReceivedServerCount` be `true` or `false` by default?
My take: have it `true` by default unless an alternative server count method is provided in the commandline args to the proxy agent. This ensures backwards compatibility with old proxy servers.

### CODE REVIEW

**Comment by carreter** [comment]
@avrittrohwer for visibility

File: pkg/agent/clientset.go [hunk 1]
```diff
@@ -63,6 +64,8 @@ type ClientSet struct {
 	warnOnChannelLimit bool
 
 	syncForever bool // Continue syncing (support dynamic server count).
+
+	respectReceivedServerCount bool // Respect server count received from proxy server rather than relying on the agent's own server counter
```
Comment (cheftako) [inline pkg/agent/clientset.go:68]: Should chat on this. During certain events with multiple KASs/Konnectivity servers we can get different server counts from different servers.

Comment (carreter) [reply pkg/agent/clientset.go:68]: Yeah, @avrittrohwer also raised this issue. One potential solution would be to cache the last n received server counts and take the highest value. Ultimately this gets solved by having proxy agents count the leases themselves.

**Pull request closed without merging**